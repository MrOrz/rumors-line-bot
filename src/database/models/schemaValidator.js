import Ajv from 'ajv';

const ajv = new Ajv();

const SCHEMAS = {
  userArticleLink: require('./userArticleLink.json'),
  userSettings: require('./userSettings.json'),
};

const ClassTable = {
  Date: Date,
};

ajv.addKeyword('instanceOf', {
  compile: function(className) {
    var Class = ClassTable[className];
    return function(data) {
      return data instanceof Class;
    };
  },
});

export function validate(schemaName, data) {
  const valid = ajv.validate(SCHEMAS[schemaName], data);
  return { valid, errors: ajv.errors };
}

export function compile(schemaName) {
  return data => {
    const validate = ajv.compile(SCHEMAS[schemaName]);
    const valid = validate(data);
    return { valid, errors: validate.errors };
  };
}
