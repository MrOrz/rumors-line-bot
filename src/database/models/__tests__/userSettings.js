import Client from '../../mongoClient';
import MockDate from 'mockdate';
import { validators } from '../schemaValidator';
import UserSettings from '../userSettings';

const userSettingsValidator = validators.userSettings;

describe('userSettings', () => {
  beforeAll(async () => {
    MockDate.set(612921600000);

    if (await UserSettings.collectionExists()) {
      await (await UserSettings.client).drop();
    }
  });

  afterAll(async () => {
    await (await Client.getInstance()).close();
  });

  it('[schema] should pass with newReplyNotifyToken is null', async () => {
    const data = {
      userId: 'this_is_user_id',
      allowNewReplyUpdate: true,
      newReplyNotifyToken: null,
    };
    const result = userSettingsValidator(data);
    expect(result).toMatchSnapshot();
  });

  it('[schema] should pass with newReplyNotifyToken is string', async () => {
    const data = {
      userId: 'this_is_user_id',
      allowNewReplyUpdate: true,
      newReplyNotifyToken: 'this_is_token',
    };
    const result = userSettingsValidator(data);
    expect(result).toMatchSnapshot();
  });

  it('[schema] should fail with missing userId', async () => {
    const data = {
      allowNewReplyUpdate: true,
      newReplyNotifyToken: null,
    };
    const result = userSettingsValidator(data);
    expect(result).toMatchSnapshot();
  });

  it('[model] should create a document', async () => {
    const result = await UserSettings.create({ userId: 'this_is_user_id' });
    delete result._id;

    expect(result).toMatchSnapshot();
  });

  it('[model] should fail to create a document', async () => {
    expect(
      UserSettings.create({
        allowNewReplyUpdate: false,
      })
    ).rejects.toThrowErrorMatchingSnapshot();
  });

  it('[model] findByUserId()', async () => {
    const userId = 'userId-0';
    await UserSettings.create({ userId, allowNewReplyUpdate: false });

    const result = await UserSettings.findByUserId(userId);
    delete result._id;

    expect(result).toMatchSnapshot();
  });

  it('[model] findByUserId() upsert', async () => {
    const userId = 'userId-1';
    const result = await UserSettings.findByUserId(userId);
    delete result._id;

    expect(result).toMatchSnapshot();
  });

  it('[model] setAllowNewReplyUpdate()', async () => {
    const userId = 'userId-2';
    await UserSettings.create({ userId, allowNewReplyUpdate: true });

    const result = await UserSettings.setAllowNewReplyUpdate(userId, false);

    delete result._id;

    expect(result).toMatchSnapshot();
  });

  it('[model] setAllowNewReplyUpdate() upsert', async () => {
    const userId = 'userId-3';
    const result = await UserSettings.setAllowNewReplyUpdate(userId, false);

    delete result._id;

    expect(result).toMatchSnapshot();
  });

  it('[model] setNewReplyNotifyToken()', async () => {
    const userId = 'userId-4';
    await UserSettings.create({ userId, newReplyNotifyToken: 'notify-token' });

    const result = await UserSettings.setNewReplyNotifyToken(
      userId,
      'notify-token-new'
    );

    delete result._id;

    expect(result).toMatchSnapshot();
  });

  it('[model] setNewReplyNotifyToken() upsert', async () => {
    const userId = 'userId-5';
    const result = await UserSettings.setNewReplyNotifyToken(
      userId,
      'notify-token'
    );

    delete result._id;

    expect(result).toMatchSnapshot();
  });
});
