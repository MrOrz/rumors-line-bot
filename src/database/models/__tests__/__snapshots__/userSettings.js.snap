// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`userSettings [model] findByUserId() 1`] = `
Object {
  "allowNewReplyUpdate": false,
  "createdAt": 1989-06-04T00:00:00.000Z,
  "newReplyNotifyToken": null,
  "userId": "userId-0",
}
`;

exports[`userSettings [model] findByUserId() upsert 1`] = `
Object {
  "allowNewReplyUpdate": true,
  "createdAt": 1989-06-04T00:00:00.000Z,
  "newReplyNotifyToken": null,
  "userId": "userId-1",
}
`;

exports[`userSettings [model] setAllowNewReplyUpdate() 1`] = `
Object {
  "allowNewReplyUpdate": false,
  "createdAt": 1989-06-04T00:00:00.000Z,
  "newReplyNotifyToken": null,
  "userId": "userId-2",
}
`;

exports[`userSettings [model] setAllowNewReplyUpdate() upsert 1`] = `
Object {
  "allowNewReplyUpdate": false,
  "createdAt": 1989-06-04T00:00:00.000Z,
  "newReplyNotifyToken": null,
  "userId": "userId-3",
}
`;

exports[`userSettings [model] setNewReplyNotifyToken() 1`] = `
Object {
  "allowNewReplyUpdate": true,
  "createdAt": 1989-06-04T00:00:00.000Z,
  "newReplyNotifyToken": "notify-token-new",
  "userId": "userId-4",
}
`;

exports[`userSettings [model] setNewReplyNotifyToken() upsert 1`] = `
Object {
  "allowNewReplyUpdate": true,
  "createdAt": 1989-06-04T00:00:00.000Z,
  "newReplyNotifyToken": "notify-token",
  "userId": "userId-5",
}
`;

exports[`userSettings [model] should create a document 1`] = `
Object {
  "allowNewReplyUpdate": true,
  "createdAt": 1989-06-04T00:00:00.000Z,
  "newReplyNotifyToken": null,
  "userId": "this_is_user_id",
}
`;

exports[`userSettings [model] should fail to create a document 1`] = `
"[
  {
    \\"keyword\\": \\"required\\",
    \\"dataPath\\": \\"\\",
    \\"schemaPath\\": \\"#/required\\",
    \\"params\\": {
      \\"missingProperty\\": \\"userId\\"
    },
    \\"message\\": \\"should have required property 'userId'\\"
  }
]"
`;

exports[`userSettings [schema] should fail with missing userId 1`] = `
Object {
  "errors": Array [
    Object {
      "dataPath": "",
      "keyword": "required",
      "message": "should have required property 'userId'",
      "params": Object {
        "missingProperty": "userId",
      },
      "schemaPath": "#/required",
    },
  ],
  "valid": false,
}
`;

exports[`userSettings [schema] should pass with newReplyNotifyToken is null 1`] = `
Object {
  "errors": null,
  "valid": true,
}
`;

exports[`userSettings [schema] should pass with newReplyNotifyToken is string 1`] = `
Object {
  "errors": null,
  "valid": true,
}
`;
